src/nupic/experimental/ExtendedTemporalMemory.cpp

```diff
@@ -0,0 +1,1316 @@
+        }
+        current_.activeSegmentsEnd = activeSegment_;
+
+        current_.matchingSegmentsBegin = matchingSegment_;
+        while (matchingSegment_ != matchingSegmentsEnd_ &&
+               columnOf_(*matchingSegment_) == current_.column)
+        {
+          matchingSegment_++;
+        }
+        current_.matchingSegmentsEnd = matchingSegment_;
+      }
+      else
+      {
+        finished_ = true;
+      }
+    }
+
+    vector<UInt>::const_iterator activeColumn_;
+    vector<UInt>::const_iterator activeColumnsEnd_;
+    vector<SegmentOverlap>::const_iterator activeSegment_;
+    vector<SegmentOverlap>::const_iterator activeSegmentsEnd_;
+    vector<SegmentOverlap>::const_iterator matchingSegment_;
+    vector<SegmentOverlap>::const_iterator matchingSegmentsEnd_;
+    const UInt cellsPerColumn_;
+
+    bool finished_;
+    ExcitedColumnData current_;
+  };
+
+  Iterator begin()
+  {
+    return Iterator(activeColumns_.begin(),
+                    activeColumns_.end(),
+                    activeSegments_.begin(),
+                    activeSegments_.end(),
+                    matchingSegments_.begin(),
+                    matchingSegments_.end(),
+                    cellsPerColumn_);
+  }
+
+  Iterator end()
+  {
+    return Iterator(activeColumns_.end(),
+                    activeColumns_.end(),
+                    activeSegments_.end(),
+                    activeSegments_.end(),
+                    matchingSegments_.end(),
+                    matchingSegments_.end(),
+                    cellsPerColumn_);
+  }
+
+private:
+  const vector<UInt>& activeColumns_;
+  const UInt cellsPerColumn_;
+  const vector<SegmentOverlap>& activeSegments_;
+  const vector<SegmentOverlap>& matchingSegments_;
+};
+
+static CellIdx getLeastUsedCell(
+  Connections& connections,
+  Random& rng,
+  UInt column,
+  UInt cellsPerColumn)
+{
+  vector<CellIdx> leastUsedCells;
+  UInt32 minNumSegments = UINT_MAX;
+  const CellIdx start = column * cellsPerColumn;
+  const CellIdx end = start + cellsPerColumn;
+  for (CellIdx cell = start; cell < end; cell++)
+  {
+    UInt32 numSegments = connections.segmentsForCell(cell).size();
+
+    if (numSegments < minNumSegments)
+    {
+      minNumSegments = numSegments;
+      leastUsedCells.clear();
+    }
+
+    if (numSegments == minNumSegments)
+    {
+      leastUsedCells.push_back(cell);
+    }
+  }
+
+  return leastUsedCells[rng.getUInt32(leastUsedCells.size())];
+}
+
+static void adaptSegment(
+  Connections& connections,
+  Segment segment,
+  const vector<CellIdx>& prevActiveInternalCells,
+  const vector<CellIdx>& prevActiveExternalCells,
+  Permanence permanenceIncrement,
+  Permanence permanenceDecrement)
+{
+  vector<Synapse> synapses = connections.synapsesForSegment(segment);
+
+  for (Synapse synapse : synapses)
+  {
+    const SynapseData synapseData = connections.dataForSynapse(synapse);
+
+    const bool isActive = (synapseData.presynapticCell < connections.numCells())
```

Comment:
Ok. I agree the ternary operator is a little obnoxious, but I think "const" makes code easier to read.